File: cmd/agenthub-gui/frontend/src/pages/profiles.ts
// Profiles page: the named tier of the scope chain.
//
// Two three-state selectors live here and both have the same failure mode:
// the difference between "no rule" and "an empty allow list" is the
// difference between exposing everything and exposing nothing. The UI keeps
// them as three visible, separately-named states — never as a list that
// happens to be empty — and refuses "only these" with nothing ticked instead
// of sending it (api/profiles.go, ui.triState).
//
// A rename is an operation, not a delete-then-create: the daemon repoints
// every client that referenced the old name, and reports which ones. A delete
// deliberately does NOT rewrite references — the clients left behind resolve
// to an EMPTY scope, which is the fail-closed direction, and they come back
// in the write's `dangling` list so the page can say it out loud.

import { hub, knownTools } from "../bridge";
import { clear, el, empty, pageHeader } from "../dom";
import type { Page } from "../page";
import { failureBox, noticeSlot, runWrite } from "../page";
import {
  button,
  confirmAction,
  controls,
  describeSelector,
  describeServerSet,
  field,
  formHost,
  selectInput,
  textInput,
  triState,
} from "../ui";
import type { Profile, ProfileList, ProfileTools, Server } from "../types";
import { ServerSet, ToolSelect } from "../types";

/** Maps a three-state selection onto the member-server field: null clears the
 *  narrowing, [] is block-all, [...] is exactly those. */
function serversFrom(sel: ProfileTools): string[] | null {
  if (sel.mode === ToolSelect.All) return null;
  if (sel.mode === ToolSelect.None) return [];
  return sel.tools ?? [];
}

function describeServersChoice(servers: string[] | null): string {
  if (servers === null) return "every registered server";
  if (servers.length === 0) return "no server at all";
  return servers.join(", ");
}

export function profilesPage(): Page {
  let root: HTMLElement | null = null;
  const slot = noticeSlot();
  const form = formHost();
  let list: ProfileList | null = null;
  let servers: Server[] = [];

  const serverIds = (): string[] => servers.map((s) => s.id);
  const generation = (): number => list?.generation ?? 0;

  // -- creation -------------------------------------------------------------

  function createForm(): Node {
    const name = textInput("", "profile name");
    const members = triState(serverIds(), undefined, {
      all: "Every registered server",
      only: "Only the ticked servers",
      none: "No server at all (block-all)",
    });
    const errors = el("div", { class: "notice-slot" });
    const save = button("Create profile", "btn", () => {
      clear(errors);
      if (!name.value.trim()) {
        errors.append(el("div", { class: "notice notice-warn", text: "A profile needs a name." }));
        return;
      }
      const sel = members.value();
      if (!sel.ok) {
        errors.append(el("div", { class: "notice notice-warn", text: sel.message }));
        return;
      }
      const chosen = serversFrom(sel.selection);
      void runWrite(
        slot,
        () => draw(),
        (r) => `Profile ${r.name} created (${describeServersChoice(chosen)}).`,
        () => hub.createProfile(name.value.trim(), chosen, generation()),
      ).then((ok) => {
        if (ok) form.hide();
      });
    });
    return el("div", { class: "panel panel-inset" }, [
      el("h3", { text: "New profile" }),
      errors,
      field("Name", name),
      field("Member servers", members.node),
      controls(save, button("Cancel", "btn btn-secondary", () => form.hide())),
    ]);
  }

  // -- member set -----------------------------------------------------------

  function membersForm(p: Profile): Node {
    const members = triState(serverIds(), { allow: p.servers }, {
      all: "Every registered server",
      only: "Only the ticked servers",
      none: "No server at all (block-all)",
    });
    const errors = el("div", { class: "notice-slot" });
    const save = button("Replace member set", "btn", () => {
      clear(errors);
      const sel = members.value();
      if (!sel.ok) {
        errors.append(el("div", { class: "notice notice-warn", text: sel.message }));
        return;
      }
      const chosen = serversFrom(sel.selection);
      void runWrite(
        slot,
        () => draw(),
        (r) => `${r.name}: members are now ${describeServersChoice(chosen)}.`,
        () =>
          hub.setProfileServers(
            p.name,
            { mode: ServerSet.Replace, servers: chosen },
            generation(),
          ),
      ).then((ok) => {
        if (ok) form.hide();
      });
    });
    return el("div", { class: "panel panel-inset" }, [
      el("h3", { text: `Member servers of ${p.name}` }),
      errors,
      members.node,
      el("p", {
        class: "hint",
        text: "“No server at all” is stored as an explicit empty set. It is not the same thing as having no rule, which exposes everything.",
      }),
      controls(save, button("Cancel", "btn btn-secondary", () => form.hide())),
    ]);
  }

  // -- per-server tool selector ---------------------------------------------

  async function toolsForm(p: Profile, server: string): Promise<Node> {
    const available = await knownTools(server);
    const picker = triState(available, p.tools?.[server]);
    const errors = el("div", { class: "notice-slot" });
    const save = button("Apply selector", "btn", () => {
      clear(errors);
      const sel = picker.value();
      if (!sel.ok) {
        errors.append(el("div", { class: "notice notice-warn", text: sel.message }));
        return;
      }
      void runWrite(
        slot,
        () => draw(),
        (r) => `${r.name}: ${server} is now "${sel.selection.mode}".`,
        () => hub.setProfileTools(p.name, server, sel.selection, generation()),
      ).then((ok) => {
        if (ok) form.hide();
      });
    });
    return el("div", { class: "panel panel-inset" }, [
      el("h3", { text: `Tools of ${server} in ${p.name}` }),
      errors,
      picker.node,
      el("p", {
        class: "hint",
        text: "Selectors are keyed by the ORIGINAL downstream tool names, not by any renamed exposure — otherwise a rename would walk out from under its own narrowing rule.",
      }),
      controls(save, button("Cancel", "btn btn-secondary", () => form.hide())),
    ]);
  }

  function toolRuleChooser(p: Profile): Node {
    const server = selectInput(
      serverIds().map((id) => ({ value: id, label: id })),
      serverIds()[0] ?? "",
    );
    const open = button("Edit selector", "btn", () => {
      if (!server.value) return;
      void toolsForm(p, server.value).then((node) => form.show(node));
    });
    return el("div", { class: "panel panel-inset" }, [
      el("h3", { text: `Tool rules of ${p.name}` }),
      field("Server", server, "the rule applies to this server's tools inside this profile"),
      controls(open, button("Cancel", "btn btn-secondary", () => form.hide())),
    ]);
  }

  // -- destructive / naming -------------------------------------------------

  function rename(p: Profile): void {
    const input = textInput(p.name, "new name");
    const errors = el("div", { class: "notice-slot" });
    form.show(
      el("div", { class: "panel panel-inset" }, [
        el("h3", { text: `Rename ${p.name}` }),
        errors,
        field("New name", input, "every client and project reference is repointed automatically"),
        controls(
          button("Rename", "btn", () => {
            clear(errors);
            const next = input.value.trim();
            if (!next || next === p.name) {
              errors.append(
                el("div", { class: "notice notice-warn", text: "Enter a different name." }),
              );
              return;
            }
            void runWrite(
              slot,
              () => draw(),
              (r) =>
                `${r.old_name ?? p.name} renamed to ${r.name}` +
                (r.repointed?.length ? ` (repointed: ${r.repointed.join(", ")})` : "") +
                ".",
              () => hub.renameProfile(p.name, next, generation()),
            ).then((ok) => {
              if (ok) form.hide();
            });
          }),
          button("Cancel", "btn btn-secondary", () => form.hide()),
        ),
      ]),
    );
  }

  async function remove(p: Profile): Promise<void> {
    const ok = await confirmAction({
      title: `Delete profile ${p.name}?`,
      body: "The profile and its selectors are removed.",
      consequences: [
        "Clients still bound to it are NOT rewritten: each one resolves to an empty scope until it is rebound, which is the fail-closed direction.",
        "The affected client ids are reported after the delete.",
      ],
      confirmLabel: "Delete",
      danger: true,
    });
    if (!ok) return;
    await runWrite(
      slot,
      () => draw(),
      (r) =>
        `Profile ${r.name} deleted.` +
        (r.dangling?.length
          ? ` These clients now resolve to an EMPTY scope: ${r.dangling.join(", ")}.`
          : "") +
        (r.active_cleared ? " The active-profile marker was cleared." : ""),
      () => hub.deleteProfile(p.name, generation()),
    );
  }

  async function setActive(p: Profile, active: boolean): Promise<void> {
    if (!active) {
      const ok = await confirmAction({
        title: "Clear the active profile?",
        body: "Every client that does not name a profile itself falls back to seeing every registered server.",
        confirmLabel: "Clear",
        danger: true,
      });
      if (!ok) return;
    }
    await runWrite(
      slot,
      () => draw(),
      () => (active ? `${p.name} is now the active profile.` : "The active-profile marker is cleared."),
      () =>
        active
          ? hub.setActiveProfile(p.name, generation())
          : hub.clearActiveProfile(p.name, generation()),
    );
  }

  // -- rendering ------------------------------------------------------------

  function toolsCell(p: Profile): Node {
    const entries = Object.entries(p.tools ?? {});
    if (entries.length === 0) return el("span", { class: "muted", text: "no rule (all tools)" });
    return el(
      "div",
      { class: "rows" },
      entries.map(([server, sel]) =>
        el("div", { class: "install" }, [
          el("strong", { text: server }),
          el("span", { class: sel.allow?.length === 0 ? "badge badge-unhealthy" : "muted", text: describeSelector(sel) }),
        ]),
      ),
    );
  }

  function profileCard(p: Profile, active: string): Node {
    const isActive = active === p.name;
    return el("article", { class: `access-card${isActive ? " access-card-active" : ""}` }, [
      el("div", { class: "access-card-head" }, [
        el("div", {}, [
          el("div", { class: "access-title-row" }, [
            el("strong", { class: "access-title", text: p.name }),
            isActive ? el("span", { class: "badge badge-healthy", text: "active default" }) : null,
          ]),
          el("div", {
            class: "muted",
            text: isActive
              ? "Used by clients that do not select a profile themselves."
              : "A reusable access boundary for one or more clients.",
          }),
        ]),
        controls(
          isActive
            ? button("Clear default", "btn btn-secondary", () => void setActive(p, false))
            : button("Make default", "btn btn-secondary", () => void setActive(p, true)),
          button("Rename", "btn btn-secondary", () => void rename(p)),
          button("Delete", "btn btn-deny", () => void remove(p)),
        ),
      ]),
      el("div", { class: "access-card-grid" }, [
        el("section", { class: "access-rule" }, [
          el("div", { class: "access-rule-head" }, [
            el("span", { class: "access-rule-label", text: "Member servers" }),
            button("Edit members", "btn btn-secondary", () => form.show(membersForm(p))),
          ]),
          el("div", {
            class: p.servers?.length === 0 ? "access-rule-value danger" : "access-rule-value",
            text: describeServerSet(p.servers),
          }),
          el("span", { class: "hint", text: "This is the first intersection in the profile scope." }),
        ]),
        el("section", { class: "access-rule" }, [
          el("div", { class: "access-rule-head" }, [
            el("span", { class: "access-rule-label", text: "Tool selectors" }),
            button("Edit rules", "btn btn-secondary", () => form.show(toolRuleChooser(p))),
          ]),
          toolsCell(p),
        ]),
      ]),
    ]);
  }

  async function draw(): Promise<void> {
    if (!root) return;
    try {
      // A Go nil slice marshals to null, so every list is normalised on
      // arrival: a page that throws on null renders nothing at all, which
      // looks exactly like "there is no configuration".
      const [pl, sl] = await Promise.all([hub.listProfiles(), hub.listServers()]);
      list = { ...pl, profiles: pl.profiles ?? [] };
      servers = sl ?? [];
    } catch (err) {
      clear(root);
      root.append(failureBox(err));
      return;
    }
    clear(root);
    const active = list.active;
    root.append(
      pageHeader(
        "Profiles",
        "Define reusable server and tool boundaries, then assign clients to the right one.",
        button("New profile", "btn btn-primary", () => form.show(createForm())),
      ),
      slot.node,
      form.node,
      el("div", { class: "scope-note" }, [
        el("span", { class: "scope-note-mark", text: "∩" }),
        el("span", {
          text: "Profiles can only narrow what a server already exposes. An empty selection is deliberately block-all, never “no rule”.",
        }),
      ]),
      list.profiles.length === 0
        ? empty("No profiles configured. Without one, every client sees every registered server.")
        : el("div", { class: "access-card-list" }, list.profiles.map((p) => profileCard(p, active))),
      el("p", {
        class: "hint page-footnote",
        text: list.active_known
          ? active
            ? `Clients that do not name a profile follow “${active}”.`
            : "No default profile: clients that do not name one see every registered server."
          : "This daemon cannot report the default profile at all — which is not the same as there being none.",
      }),
    );
  }

  return {
    render(node) {
      root = node;
      return draw();
    },
    dispose() {
      root = null;
      form.hide();
    },
  };
}

File: api/profiles.go
package api

import (
	"context"
	"net/http"
	"net/url"
)

// Tool selector modes (docs/model.md). The three states are keyed by
// ORIGINAL tool names — never exposed/renamed names, otherwise a rename
// would walk out from under its own narrowing rule.
//
// WHY AN EXPLICIT MODE AND NOT A BARE LIST: the three states are "every
// tool", "exactly these" and "no tool at all", and the last two differ ONLY
// by an empty vs non-empty list. Any encoding in which a missing field and
// an empty list can be confused collapses block-all into allow-all the first
// time a marshaller drops an empty slice, an HTTP layer normalizes a body or
// a hand-written client forgets the field. Spelling the state out makes that
// class of bug unrepresentable: a caller who omits Mode sends "" and is
// REFUSED, instead of silently opening a server up.
const (
	// ToolSelectAll exposes the server's full tool set. At a single layer
	// that is indistinguishable from "no rule", so the rule is dropped
	// rather than stored as an inert object.
	ToolSelectAll = "all"
	// ToolSelectOnly narrows to the named subset (at least one name).
	ToolSelectOnly = "only"
	// ToolSelectNone blocks every tool of the server: the EMPTY allow list.
	// It is stored, not dropped — dropping it would flip block-all into
	// allow-all.
	ToolSelectNone = "none"
)

// ProfileTools is one three-state tool selector EDIT.
//
// Failure direction: an empty Mode is the zero value and is REFUSED by the
// daemon. Making the zero value mean ToolSelectAll would let a caller that
// forgot the field silently widen a selector; unset must never be the loose
// case.
type ProfileTools struct {
	// Mode is one of the ToolSelect* constants. Required — no omitempty, so
	// a forgotten mode reaches the daemon as "" and is refused there rather
	// than being guessed here.
	Mode string `json:"mode"`
	// Tools are the RAW downstream tool names, required by ToolSelectOnly
	// and ignored otherwise. An empty Tools with mode "only" is NOT how
	// "none" is expressed: it is a caller bug and is refused, because
	// guessing would pick the fail-open reading.
	Tools []string `json:"tools,omitempty"`
}

// AllTools drops the tool rule for a server (its full set is exposed).
func AllTools() ProfileTools { return ProfileTools{Mode: ToolSelectAll} }

// OnlyTools narrows a server to the named tools.
func OnlyTools(tools ...string) ProfileTools {
	return ProfileTools{Mode: ToolSelectOnly, Tools: tools}
}

// NoTools blocks every tool of a server. Using this constructor rather than
// an empty subset is what keeps block-all spellable and unmistakable.
func NoTools() ProfileTools { return ProfileTools{Mode: ToolSelectNone} }

// ToolSelector is a STORED three-state selector as read back. It is the
// registry's own shape, so the same nil-vs-empty distinction the daemon
// keeps on disk survives the round trip:
//
//	Allow == nil  -> the server's full tool set (no rule)
//	Allow == []   -> block-all
//	Allow == [..] -> exactly those tools
//
// `omitzero` (not omitempty) is load-bearing: it keeps the empty list on the
// wire, and dropping it would silently turn block-all into allow-all.
type ToolSelector struct {
	Allow []string `json:"allow,omitzero"`
}

// Blocked reports the block-all state: an allow list that is present and
// empty. It exists so a frontend never has to write the nil-vs-empty test
// itself — the one test whose wrong answer opens a server up.
func (s ToolSelector) Blocked() bool { return s.Allow != nil && len(s.Allow) == 0 }

// Profile is one named tier of the scope chain as READ from the daemon.
type Profile struct {
	Name string `json:"name"`
	// Servers is the three-state member set, with the same rule as
	// ToolSelector.Allow: absent = no narrowing (every registered server),
	// [] = block-all, [...] = exactly those. `omitzero` keeps the empty
	// list distinguishable from the absent one after a decode.
	Servers []string `json:"servers,omitzero"`
	// Tools maps a server id to its stored selector.
	Tools map[string]ToolSelector `json:"tools,omitempty"`
}

// Blocked reports that this profile's member set is present and empty, i.e.
// it exposes no server at all.
func (p Profile) Blocked() bool { return p.Servers != nil && len(p.Servers) == 0 }

// ProfileList is the answer to Profiles.List.
type ProfileList struct {
	// Generation is what the following write sends as expectedGeneration.
	Generation uint64    `json:"generation"`
	Profiles   []Profile `json:"profiles"`
	// Active is the globally active profile ("" = none).
	Active string `json:"active"`
	// ActiveKnown is false when this daemon cannot answer the question at
	// all (no state directory). A frontend must tell "there is no active
	// profile" from "this daemon does not know" — the two look identical in
	// Active alone.
	ActiveKnown bool `json:"active_known"`
}

// Server-set edit modes for ServerSetEdit.Mode.
const (
	// ServerSetReplace sets the list outright: null clears the narrowing
	// (every registered server), [] is block-all.
	ServerSetReplace = "replace"
	// ServerSetAdd adds ids. A profile with NO narrowing becomes an
	// explicit set the moment one server is named: "these and only these".
	ServerSetAdd = "add"
	// ServerSetRemove drops ids from the list.
	ServerSetRemove = "remove"
)

// ServerSetEdit is one edit of a profile's member-server set. The
// read-modify-write for add/remove happens under the registry lock, so a
// concurrent edit cannot be lost the way a caller computing the new list
// from a stale snapshot would lose it.
type ServerSetEdit struct {
	// Mode is one of the ServerSet* constants. Required; "" is refused
	// rather than defaulted.
	Mode string `json:"mode"`
	// Servers is the id list the mode applies to. Under ServerSetReplace an
	// explicit empty list is block-all and a null is "no narrowing", so this
	// field has no omitempty: the empty list has to survive the wire.
	Servers []string `json:"servers"`
}

// ProfileToolsEdit scopes one three-state selector to one server, which is
// exactly the shape of the operation behind it.
type ProfileToolsEdit struct {
	Server string `json:"server"`
	ProfileTools
}

// ProfilePatch is an edit of one profile. EXACTLY ONE field may be set: they
// are separate operations with no transaction spanning them, so a combined
// request could half-apply. The daemon refuses anything else.
type ProfilePatch struct {
	// Rename renames the profile AND repoints every client and project
	// reference to it. A rename is an operation, not a delete-then-create:
	// leaving references behind would fail-close every one of those clients
	// to an EMPTY scope.
	Rename string `json:"rename,omitempty"`
	// Servers edits the membership set.
	Servers *ServerSetEdit `json:"servers,omitempty"`
	// Tools sets one server's three-state tool selector.
	Tools *ProfileToolsEdit `json:"tools,omitempty"`
	// Active points the global active marker: true points it at THIS
	// profile, false clears it (there is no "some other profile" to guess
	// at). A nil pointer leaves the marker alone — which is why this is a
	// pointer: false must mean "clear", never "did not mention".
	Active *bool `json:"active,omitempty"`
}

// ProfileWrite is the answer to every profile mutation.
type ProfileWrite struct {
	WriteResult
	Name    string `json:"name"`
	OldName string `json:"old_name,omitempty"`
	// Profile is the entry as it now stands (absent after a delete).
	Profile *Profile `json:"profile,omitempty"`
	// Repointed lists the client ids a rename rewrote.
	Repointed []string `json:"repointed,omitempty"`
	// Dangling lists client ids left pointing at a REMOVED profile. They are
	// not rewritten (fail-closed: they now resolve to an EMPTY scope); they
	// are reported so a frontend can say so out loud.
	Dangling      []string `json:"dangling,omitempty"`
	ActiveCleared bool     `json:"active_cleared,omitempty"`
	Deleted       bool     `json:"deleted,omitempty"`
}

// profileCreateBody is the POST /v1/profiles body.
type profileCreateBody struct {
	Name string `json:"name"`
	// Servers is the initial membership. The pointer preserves the three
	// states: nil omits the key (no narrowing), &[]string{} sends [] which
	// is block-all. A plain slice with omitempty would encode those two
	// identically and collapse block-all into allow-all.
	Servers *[]string `json:"servers,omitempty"`
}

// ProfilesService manages the profile tier of the scope chain.
type ProfilesService struct{ c *Client }

// List returns every configured profile plus the active marker.
func (s *ProfilesService) List(ctx context.Context) (ProfileList, error) {
	var out ProfileList
	err := s.c.do(ctx, http.MethodGet, "/profiles", nil, nil, &out)
	return out, err
}

// Create adds a profile. servers keeps the three-state distinction: nil
// creates a profile that sees every registered server, a pointer to an empty
// slice creates a block-all one. The two are never collapsed.
func (s *ProfilesService) Create(
	ctx context.Context, name string, servers *[]string, expectedGeneration uint64,
) (ProfileWrite, error) {
	var out ProfileWrite
	err := s.c.doWrite(ctx, http.MethodPost, "/profiles", nil, expectedGeneration,
		profileCreateBody{Name: name, Servers: servers}, &out)
	return out, err
}

// Update applies one patch to one profile. Exactly one field of patch may be
// set.
func (s *ProfilesService) Update(
	ctx context.Context, name string, patch ProfilePatch, expectedGeneration uint64,
) (ProfileWrite, error) {
	var out ProfileWrite
	err := s.c.doWrite(ctx, http.MethodPatch, "/profiles/"+url.PathEscape(name), nil,
		expectedGeneration, patch, &out)
	return out, err
}

// Rename renames a profile and repoints every reference to it.
func (s *ProfilesService) Rename(
	ctx context.Context, name, newName string, expectedGeneration uint64,
) (ProfileWrite, error) {
	return s.Update(ctx, name, ProfilePatch{Rename: newName}, expectedGeneration)
}

// SetServers edits a profile's member-server set.
func (s *ProfilesService) SetServers(
	ctx context.Context, name string, edit ServerSetEdit, expectedGeneration uint64,
) (ProfileWrite, error) {
	return s.Update(ctx, name, ProfilePatch{Servers: &edit}, expectedGeneration)
}

// SetTools sets one server's three-state tool selector inside a profile.
// Both the profile and the server must exist: a selector naming a server the
// registry does not know narrows nothing, so accepting it would be a rule
// that silently does not apply.
func (s *ProfilesService) SetTools(
	ctx context.Context, name, server string, sel ProfileTools, expectedGeneration uint64,
) (ProfileWrite, error) {
	return s.Update(ctx, name,
		ProfilePatch{Tools: &ProfileToolsEdit{Server: server, ProfileTools: sel}}, expectedGeneration)
}

// Delete removes a profile.
//
// Referencing clients are deliberately NOT rewritten: a dangling reference
// resolves to an EMPTY scope, never a widened one. They come back in
// ProfileWrite.Dangling and in Warnings, because "this client just lost every
// tool" is not something an operator may learn by accident.
func (s *ProfilesService) Delete(
	ctx context.Context, name string, expectedGeneration uint64,
) (ProfileWrite, error) {
	var out ProfileWrite
	err := s.c.doWrite(ctx, http.MethodDelete, "/profiles/"+url.PathEscape(name), nil,
		expectedGeneration, nil, &out)
	return out, err
}

// SetActive points the global active marker at a profile. The profile must
// exist: the marker is the fallback for every client that does not name a
// profile itself, so pointing it at a typo would fail-close all of them at
// once.
func (s *ProfilesService) SetActive(
	ctx context.Context, name string, expectedGeneration uint64,
) (ProfileWrite, error) {
	active := true
	return s.Update(ctx, name, ProfilePatch{Active: &active}, expectedGeneration)
}

// ClearActive drops the global active marker, which currently points at
// name; every client that does not name a profile itself then sees every
// registered server again.
func (s *ProfilesService) ClearActive(
	ctx context.Context, name string, expectedGeneration uint64,
) (ProfileWrite, error) {
	active := false
	return s.Update(ctx, name, ProfilePatch{Active: &active}, expectedGeneration)
}
